toolchain/check/eval.cpp

```diff
@@ -1549,641 +1544,672 @@ static auto MakeFacetTypeResult(Context& context,
-              eval_context, &typed_inst,
-              &SemIR::FacetAccessType::facet_value_inst_id, &phase)) {
-        if (auto facet_value = eval_context.insts().TryGetAs<SemIR::FacetValue>(
-                typed_inst.facet_value_inst_id)) {
-          return eval_context.constant_values().Get(facet_value->type_inst_id);
-        }
-        return MakeConstantResult(eval_context.context(), typed_inst, phase);
-      } else {
-        return MakeNonConstantResult(phase);
-      }
+
+    LoadImportRef(context, element);
+    return ConstantEvalResult::Existing(GetConstantValueInSpecific(
+        context.sem_ir(), witness->specific_id, element));
+  }
+
+  return ConstantEvalResult::New(inst);
+}
+
+static auto EvalConstantInst(Context& /*context*/, SemIRLoc /*loc*/,
+                             SemIR::BindValue /*inst*/) -> ConstantEvalResult {
+  // TODO: Handle this once we've decided how to represent constant values of
+  // reference expressions.
+  return ConstantEvalResult::TODO;
+}
+
+static auto EvalConstantInst(Context& /*context*/, SemIRLoc /*loc*/,
+                             SemIR::Deref /*inst*/) -> ConstantEvalResult {
+  // TODO: Handle this.
+  return ConstantEvalResult::TODO;
+}
+
+static auto EvalConstantInst(Context& /*context*/, SemIRLoc /*loc*/,
+                             SemIR::Temporary /*inst*/) -> ConstantEvalResult {
+  // TODO: Handle this. Can we just return the value of `init_id`?
+  return ConstantEvalResult::TODO;
+}
+
+static auto EvalConstantInst(Context& /*context*/, SemIRLoc /*loc*/,
+                             SemIR::VtablePtr /*inst*/) -> ConstantEvalResult {
+  // TODO: Handle this.
+  return ConstantEvalResult::TODO;
+}
+
+static auto EvalConstantInst(Context& context, SemIRLoc /*loc*/,
+                             SemIR::AsCompatible inst) -> ConstantEvalResult {
+  // AsCompatible changes the type of the source instruction; its constant
+  // value, if there is one, needs to be modified to be of the same type.
+  auto value_id = context.constant_values().Get(inst.source_id);
+  CARBON_CHECK(value_id.is_constant());
+
+  auto value_inst =
+      context.insts().Get(context.constant_values().GetInstId(value_id));
+  auto phase = GetPhase(context.constant_values(),
+                        context.types().GetConstantId(inst.type_id));
+  value_inst.SetType(inst.type_id);
+
+  // Finish computing the new phase by incorporating the phases of the
+  // arguments.
+  EvalContext eval_context(context, SemIR::InstId::None);
+  auto kinds = value_inst.ArgKinds();
+  GetConstantValueForArg(eval_context, kinds.first, value_inst.arg0(), &phase);
+  GetConstantValueForArg(eval_context, kinds.second, value_inst.arg1(), &phase);
+  CARBON_CHECK(IsConstant(phase));
+
+  // We can't use `ConstantEvalResult::New` because it would use the wrong
+  // phase, so manually build a new constant.
+  return ConstantEvalResult::Existing(
+      MakeConstantResult(context, value_inst, phase));
+}
+
```

Comment:
```suggestion

// These semantic wrappers don't change the constant value.
```

Reply:
I'm planning on putting these overloads in alphabetical order in an upcoming PR; do you think this comment is important? If so, I'll duplicate it across all of the relevant overloads.

Reply:
It is mildly helpful since the code is a bit brief and mysterious otherwise.